UI draft

modgethanc opened the pull request:
relevant to #13 and #41, i've stubbed/hardcoded in some UI changes that i think helps move the client forward. most of this was just helpful for me to visually block out the client a little more, while also learning my way around urwid and the codebase in general.

specifically:
* added a tip on the main/login screen to push 'ESC' to close windows (this took me a moment to figure out on my own when i first launched the client)
* added more info to the registration menu, including noting the hardcoded password minimum
* built on the existing main game window draft:
  * stubbed out some text that prints in game window on successful login
  * added `\refresh` command, which repopulates the room info and character info panels (in the future, if those panels update dynamically, this command will be obsolete)
* minor appearance adjustments:
  * added quotation marks around speech
  * added some styling to the banner to make it visually distinct from main game area

i'm not sure if this is really PR-ready, but i feel like i'm hitting a wall with what i can meaningfully work on at the moment; i'm happy to hold these changes back until more server-side work progresses if that's the sensible thing to do!

vilmibm commented on server/tmserver/world.py:
```diff
@@ -117,6 +119,40 @@ def handle_look(cls, sender_obj, action_args):
 
+    @classmethod
+    def handle_refresh(cls, sender_obj, action_args):
+        # TODO: should the server dictate what text goes into the client's info
```
Definitely the latter. we don't want the server to ever have to think about the UI -- its only job is passing data (which sometimes might be as simple as a string of text to show or as complex as a big json pile)

vilmibm commented on server/tmserver/world.py:
```diff
@@ -34,6 +34,8 @@ def dispatch_action(cls, sender_obj, action, action_args):
             cls.handle_whisper(sender_obj, action_args)
         if action == 'look':
             cls.handle_look(sender_obj, action_args)
+        if action == 'refresh':
```
would a user ever be typing `/refresh`? i'm thinking the client will be performing this action at some polling interval, sending a message like `DATA roominfo` and getting back a payload like `DATA roominfo {...}`. Essentially, we're crudely reinventing REST for use in a bidirectional websocket connection (instead of a unidirectional HTTP setup).

As we need JSON payloads we can add handlers for other kinds of data -- like `minimap`, `map` (for maybe a big map modal), `inventory`, `topic`...

does this make sense? i can whip up the DATA handler on the server side today, i think.

modgethanc replied:
yeah, realistically the user would never actually need to manually `/refresh`, that's mostly a debugging/instant gratification thing at the moment. but, i wonder if instead of having the client regularly poll, we just have the server update the client whenever there's actually a relevant change (new object enters the room, or you move into a different room, or the room name/description gets changed, etc.), and the client refreshes the panels when it gets fresh DATA. it seems like less work on both ends (no need to transmit DATA payloads if they haven't changed, not missing information if something happens and then unhappens within the polling interval).

modgethanc commented:
i'm definitely on board with a json payload, like basically the entire game state at any moment can theoretically be expressed that way, right? though to keep overhead minimal, maybe only transmit actually relevant info for the particular DATA request (so like an admin in god mode might need to be able to just peek into every single room and inspect every object, but a normal user doesn't necessarily need a list of objects in rooms beyond the one they're in, or within some area of effect we can define).

anyway, i think if there's a DATA handler i can use, i can make some more progress on panels!

vilmibm commented:
i think i just worded out all the things you said here 20 hours ago in irc ^_^() i think we're all on the same page, though, now!

modgethanc commented:
okay! should i hold back/revert the `\refresh`-related changes for now? there's going to be some substantial overhaul on the UI once data API and a more polished client design comes through anyway, though i think the little bits of UI polish is at least helpful for me in terms of visualizing where stuff is going.

modgethanc commented:
actually okay after poking at this more and getting some ideas from @selfsame, i'm gonna close this until it's a bit more polished. i've got a bunch of half-baked screen shuffles going on right now, so i'll get that stuff settled first!

modgethanc commented:
okay! in this updated PR:

* shows off a cleaner fresher look with nice unicode lines to delineate all the panels!
* removes the previously implemented `\refresh` command, pending better DATA api implementation
* implements a tab-based interface, with mostly stubbed tabs for now, but supports tab-switching
* autoscrolls the main game text once there are more lines than can be displayed
* includes a preliminary quit button (mapped to F9) that exits the client (though this still doesn't hook up to a clean disconnect from the server)

vilmibm commented:
:sparkles: :sparkling_heart: :sparkles: :sparkling_heart: :sparkles: :sparkling_heart: :sparkles: :sparkling_heart: :sparkles: :sparkling_heart: :sparkles: :sparkling_heart: :sparkles: :sparkling_heart: 

will review this tonight!! quick question: i noticed you're saying `\quit` instead of `/quit`. are you envisioning two kinds of slash commands? or do you think we should be using `\` instead of `/`?

modgethanc commented:
haha no i think i just got `\` in my head for some reason?? `/` is fine, i will reprogram my fingers to do that!

vilmibm commented:
aside: i love `<a None named None>`

modgethanc commented:
okay! i fixed the test case so it checks for how quoting is heard now, and also removed the keypress appearing in the lower left corner (that was leftover from when i was debugging). 

i'll make a note to add more hotkey bindings for tab switching (possibly as something that can be user-configurable later).
